pyat/at/tracking/particles.py

```diff
@@ -0,0 +1,131 @@
+"""
+Functions relating to particle generation
+"""
+import numpy
+from at.physics import ohmi_envelope
+
+__all__ = ['beam', 'sigma_matrix']
+
+
+def sigma_matrix(argin, twiss=False):
+    """
+    Calculate the correlation matrix to be used for particle generation
+
+    PARAMETERS
+        argin           Lattice object or list of twiss parameters.
+                        Depending on the length of the list the function
+                        will do different things:
+                        [espread, blength] 
+                            computes 2x2 longitudinal sigma matrix
+                        [beta, alpha, emit] 
+                            computes 2x2 transverse sigma matrix
+                        [betax, alphax, emitx, betay, alphay, emity] 
+                            computes the 4x4 sigma matrix for horizontal 
+                            and vertical.
+                        [betax, alphax, emitx, betay, alphay, emity, espread, 
+                            blength] 
+                            computes the 6x6 sigma matrix.
+
+    KEYWORDS
+        twiss=False     Flag which states whether the input is a
+                        lattice object or list of twiss parameters
+
+    OUTPUT
+        sigma_matrix    correlation matrix (either 2x2, 4x4 or 6x6)
+
+    """
+    if not twiss:
+        cflag = False
+        if not argin.radiation:
+            argin.radiation_on()
+            cflag = True
+
+        emit0, beamdata, emit = ohmi_envelope(argin, refpts=[0])
+        sig_matrix = emit.r66[0]
+
+        if cflag:
+            argin.radiation_off()
+
+
+    else:
+        if len(argin) == 2:
```

Comment:
If you rely upon the number of arguments to guess what inputs are, why not checking 
len(argin)==1
to guess that it is a Lattice object ?
Then, no need for the `twiss` flag

Reply:
If argin is a lattice object then len(argin) = nelems not 1. So I resorted to a flag. Is there a robust way to check if an argument is a lattice object?

Reply:
Ah ! I understood (wrongly) that you were expecting a variable number of arguments with the syntax `def sigma_matrix(*argin)`. Now that I see that you want a single argument which is either a Lattice object or a tuple, you could test `isinstance(argin, Lattice)`

Reply:
`isinstance(argin, at.Lattice)` returns True for a Lattice object. But it also returns positive for `isinstance(argin, list)`.

Reply:
Normal, since a Lattice is a list (but not the opposite). That's why you need to test for a Lattice, so that any other list of values like [energy_spread, bunch_length] returns False.

Reply:
ah yes, you are right. I was thinking of the reverse which is not needed. A list will return False for a lattice object check. Thanks a lot. Sorry for the confusion